certlogic/certlogic-dart/lib/src/evaluate.dart

```diff
@@ -19,21 +21,30 @@ class CertLogic {
     dynamic returnData = data;
     value.split('.').forEach((fragment) {
       if (returnData == null) return;
-      if (fragment == 'current') {
+      if (fragment == 'data') {
```

Comment:
This is out-of-spec behaviour (and same for that regarding `current` and `accumulator` below), right? Not so much of a problem (and certainly not for this PR), but I'd like to know: is there a specific reason you added it? I wouldn't mind keeping the reference implementations as close/tight to the spec as possible to avoid that people start using out-of-spec behaviour.

Reply:
According to the specs:

> The reduce operation is equivalent to a left-fold over the array <operand> with <initial> prepended, using the provided <lambda> function. That function is provided with a modified data context of the form `{ "current": <current>, "accumulator": <accumulator> }`, with the **current** equalling the items of the array (in that order), and **accumulator** equalling the result of the left-fold so far. In particular, the reduce of an empty-array operand evaluates to initial. This is essentially equivalent to JavaScript's Array.reduce function.
> 
> To be able to access values in the original data context, CertLogic may expand beyond JsonLogic at some point by also adding a key-value pair with key "data" to the data object passed to the <lambda>, whose value is the original data context.

For the data, I didn't test is fully and is indeed out-of-spec, but may be used in a future version. I've added this before noticing that we didn't skip tests that are marked with skip.

Reply:
Ah, OK: I'm starting to understand.

The JS implementation of CertLogic achieves that by actively creating a new data context of that shape for every step of the `reduce` operation. Does the Dart implementation do the same?

Reply:
How I've implemented it at this moment is a bit different. After knowing what it does I do understand the js implementation, but when I started the translation to Dart I didn't understand that part, therefore I choose another route.
We have 2 private variables (note: I will ignore data for now)
```
  static dynamic _accumulator;
  static dynamic _current;
```

In reduce, we first calculate the accumulator value and use that. Then for each value, we calculate it, store it in current and add it to accumulator using the provided lambda.
```
    _accumulator = evalInitial();
    evalOperand.forEach((dynamic accumulator) {
      _current = accumulator;
      _accumulator = _evaluate(lambda, <dynamic>[accumulator, _accumulator]);
    });
    return _accumulator;
```
Then in the var, we check if the var is "current" or "accumulator". If so, we return that value, otherwise return the normal one
```
      if (fragment == 'current') {
        returnData = _current;
        return;
      }
      if (fragment == 'accumulator') {
        returnData = _accumulator;
        return;
      }
```
So this part `<dynamic>[accumulator, _accumulator]` is actually a mess 😅. That should be a Map (data object) that provides the current and accumulator rather than a meaningless list that is never used.
I'll clean this up when I find the time 👍

Reply:
Maybe you can turn this conversation into an issue (and self-assign?), so we don't forget.